### PR TITLE
[ticket/14733] Support increasing hashing cost factor

### PR DESCRIPTION
**Pull request by marc1706**
This will need another pull request for 3.2.x.

Checklist:
- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14733

PHPBB3-14733

### CODE REVIEW

File: phpBB/phpbb/passwords/driver/driver_interface.php [hunk 1]
```diff
@@ -30,6 +30,14 @@ public function is_supported();
 	public function is_legacy();
 
 	/**
+	 * Check if password needs to be rehashed
+	 *
+	 * @param string $hash Hash to check for rehash
+	 * @return bool True if password needs to be rehashed, false if not
+	 */
+	public function needs_rehash($hash);
```
Comment (bantu) [inline phpBB/phpbb/passwords/driver/driver_interface.php]: Breaks other existing interface implementations, thus can not go into 3.1.x?

**Comment by bantu** [comment]
This can not go into 3.1.x because of https://github.com/phpbb/phpbb/pull/4405#discussion_r75744572. Additionally, it should not go into a stable branch because it is basically a new feature (with the usual possibility of regression/breaks).

File: phpBB/phpbb/passwords/driver/driver_interface.php (hunk 1 above)
Comment (marc1706) [inline phpBB/phpbb/passwords/driver/driver_interface.php]: As all drivers are extending the base class this shouldn't cause any BC breaks.

**Comment by marc1706** [comment]
I'm ok with moving this to 3.2.x as target. It shouldn't cause any BC breaks as it's using the abstract base class for an implementation of the de-facto state of the current convert flag. I do however understand that this is basically a new feature so I guess it's fair to have this moved to a 3.2.x target.

@Nicofuma agree?

**Comment by Nicofuma** [comment]
:+1: for 3.2.1 or 3.3.0 (we really need to define what we accept in each kind of release)

**Comment by bantu** [comment]
@marc1706 An extension might implement driver_interface, no?

**Comment by marc1706** [comment]
@bantu I guess you'd say 3.3.0 target in this case, i.e. master?

**Comment by bantu** [comment]
@marc1706 Yes, master, when 3.2.0 is already out.

**Comment by marc1706** [comment]
Rebased on top of master branch.

**Comment by marc1706** [comment]
@Nicofuma if you think this can go into 3.2 I'll rebase this on top of 3.2

**Comment by Nicofuma** [comment]
can be 3.2.1 as the BC issue is solved

**Comment by marc1706** [comment]
Updated base branch to 3.2.x

File: phpBB/phpbb/passwords/driver/base.php [hunk 2]
```diff
@@ -13,15 +13,15 @@
 
 namespace phpbb\passwords\driver;
 
-abstract class base implements driver_interface
+abstract class base implements rehashable_driver_interface
```
Comment (Nicofuma) [inline phpBB/phpbb/passwords/driver/base.php]: Why implementing it by default?

Comment (marc1706) [inline phpBB/phpbb/passwords/driver/base.php]: To be able to increase hashing cost factor by default. Might be used for other and future hashing types, too.

**Pull request merged**